linker.go

```diff
@@ -57,21 +62,50 @@ func splitSymbols(insns asm.Instructions) (map[string]asm.Instructions, error) {
 		return nil, errors.New("insns is empty")
 	}
 
-	if insns[0].Symbol() == "" {
+	firstSymbol := insns[0].Symbol()
+	if firstSymbol == "" {
 		return nil, errors.New("insns must start with a Symbol")
 	}
 
-	var name string
-	progs := make(map[string]asm.Instructions)
-	for _, ins := range insns {
+	// we first count the amount of symbols
+	// this is useful for pre-allocating slices later, but also to handle
+	// the fast path when we have only one symbol
+	symbolCounter := 1
+	for _, ins := range insns[1:] {
```

Comment:
Wouldn't something like the following be a bit simpler?

```diff
diff --git a/linker.go b/linker.go
index b653b805..6eeadbf1 100644
--- a/linker.go
+++ b/linker.go
@@ -6,6 +6,7 @@ import (
 	"fmt"
 	"io"
 	"math"
+	"slices"
 
 	"github.com/cilium/ebpf/asm"
 	"github.com/cilium/ebpf/btf"
@@ -57,21 +58,31 @@ func splitSymbols(insns asm.Instructions) (map[string]asm.Instructions, error) {
 		return nil, errors.New("insns is empty")
 	}
 
-	if insns[0].Symbol() == "" {
+	currentSym := insns[0].Symbol()
+	if currentSym == "" {
 		return nil, errors.New("insns must start with a Symbol")
 	}
 
-	var name string
+	start := 0
 	progs := make(map[string]asm.Instructions)
-	for _, ins := range insns {
-		if sym := ins.Symbol(); sym != "" {
-			if progs[sym] != nil {
-				return nil, fmt.Errorf("insns contains duplicate Symbol %s", sym)
-			}
-			name = sym
+	for i, ins := range insns {
+		sym := ins.Symbol()
+		if sym == "" {
+			continue
+		}
+
+		// New symbol, flush the old one out.
+		progs[currentSym] = slices.Clone(insns[start:i])
+
+		if progs[sym] != nil {
+			return nil, fmt.Errorf("insns contains duplicate Symbol %s", sym)
 		}
+		currentSym = sym
+		start = i
+	}
 
-		progs[name] = append(progs[name], ins)
+	if tail := insns[start:]; len(tail) > 0 {
+		progs[currentSym] = slices.Clone(tail)
 	}
 
 	return progs, nil
```

Completely untested, but you get the idea. No need for another type, no need to allocate indices and syms. The single symbol case would iterate through insns once and then do a copy in the `if tail` part.